### ISSUE TITLE
cmd/vet: don't report duplicate json tags when field shadowing is used

### ISSUE THREAD

**Issue by flimzy** [golang/go#30846]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.12 linux/amd64
</pre>

### Does this issue reproduce with the latest release?

Yes

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/jonhall/.cache/go-build"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH="/home/jonhall/go"
GOPROXY=""
GORACE=""
GOROOT="/usr/local/go"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GCCGO="gccgo"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build643199218=/tmp/go-build -gno-record-gcc-switches"

</pre></details>

### What did you do?

Given this simple code:

```
package test

type Foo struct {
	Foo string `json:"foo"`
	Bar string `json:"bar"`
}

type Bar struct {
	Foo
	Bar int `json:"bar"`
}
```

As of Go 1.12, `go vet` complains:

> ./test.go:10:2: struct field Bar repeats json tag "bar" also at test.go:5

### What did you expect to see?

I expect to see no complaint from `go vet`, as I did in Go 1.11 and earlier, since the field defined at line 5 is hidden, so is disregarded by the `json` package.

**Comment by mvdan** [comment]
This is https://github.com/golang/go/issues/30465; the enhancement to the vet check was temporarily reverted in 1.12.1. I plan to simply fix the check for 1.13 in master, as opposed to disabling it.

I wanted to file a separate issue to track that, but you beat me to it :) I'll use this one, if that's okay.

**Comment by dstrelau** [comment]
Hi :wave:. What is the status of putting a proper fix for this into x/tools master?

Right now, golangci/golangci-lint is in a bit of a spot where it's reporting false positives for this case, even when run on 1.12.5 because it brings in x/tools@685fecacd0a0 (master) as an explicit dependency. However, it looks like the temporary fix put in to revert this behavior was put onto the x/tools 1.12 release branch and never pulled back into master. This means that golangci-lint doesn't have the fix and can't get it — it can't simply switch to the 1.12 release branch because there are 300+ other commits in master, many of which fix other issues.

Would it be possible to pull the temporary fix from the 1.12 release branch into master for the time being, until a proper fix can be made for 1.13? I'm not sure the best course of action here.

- https://github.com/golangci/golangci-lint/blob/master/go.mod#L55
- https://github.com/golang/tools/compare/master...release-branch.go1.12

**Comment by mvdan** [comment]
Thanks for the ping @dstrelau. I wasn't in a rush to fix this, other than making it to the 1.13 release, as I didn't know anyone was running vet from x/tools master. Is there a particular reason for that, other than getting the latest features? I think that's confusing, as most developers will run vet from the latest Go version, not x/tools master.

In any case, I'll look at a fix today, as there's no reason to delay it further.

**Comment by dstrelau** [comment]
Thanks @mvdan! I don't really know why golangci-lint chose to vendor master — I'm not involved in development there, just a user — but it seems intentional as they've pulled in newer versions from master several times. I opened an issue over there, so we'll see what the maintainers say.

Either way, thanks for the speedy response and looking into this. 🥇
